# Objective:

In this challenge we do an in-depth analysis of orders at different sneaker shops at Shopify.

# Business Problem:

The average order value of sneakers from the dataset is very high i.e.  $3145.13 which is an incorrect estimate of average sales. We need to fix this problem to get a correct estimate of sales and define a more efficient metric to assess sales in this segment.

# Solution:



To get to the solution, we follow the below mentioned steps:

1. Firstly import all the libraries in Python.

2. Check the format of each variable.

3. Check for duplicate rows in the dataset.

4. Check for missing values and try imputing them if possible, if any.

5. Check for anomalies in the data using box plots.


#### Step 1: Importing all the libraries

In [4]:
import pandas as pd
import numpy as np
import seaborn as sns #visualisation
import matplotlib.pyplot as plt #visualisation
%matplotlib inline 
sns.set(color_codes=True)
import plotly.express as px # for visualisation

In [8]:
df = pd.read_csv('shopify.csv')

In [9]:
df.head()

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11

### Step 2: Checking the data type of each variable

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   order_id        5000 non-null   int64 
 1   shop_id         5000 non-null   int64 
 2   user_id         5000 non-null   int64 
 3   order_amount    5000 non-null   int64 
 4   total_items     5000 non-null   int64 
 5   payment_method  5000 non-null   object
 6   created_at      5000 non-null   object
dtypes: int64(5), object(2)
memory usage: 273.6+ KB


In [11]:
# Total number of rows and columns
df.shape


(5000, 7)

### Step 3: Check for duplicate rows in the dataset

In [12]:

# Rows containing duplicate data
duplicate_rows_df = df[df.duplicated()]
print("number of duplicate rows: ", duplicate_rows_df.shape[0])



number of duplicate rows:  0


The above analysis indicates **no duplicate rows** in the dataset.

### Step 4: Check for missing values in the dataset.

In [13]:
# Finding the null values.
print(df.isnull().sum())

order_id          0
shop_id           0
user_id           0
order_amount      0
total_items       0
payment_method    0
created_at        0
dtype: int64


**There are no missing values in the dataset.**

### Step 5: Check for outliers/anomalies in the data.

We are mainly interested in two variables here i.e. `Order Amount` and `Total number of items`. We can plot a boxplot to indicate if there any outliers in each of these two variables.

In [22]:
# Creating box plot for order amount
fig = px.box(df, y="order_amount")
fig.show()

In [23]:
#Creating box plot for number of items ordered.

fig = px.box(df, y="total_items")
fig.show()

The above two box plot indicate that:

**There is one huge outlier with respect to the order amounting to $704k.** Also there is one huge order of 2000 items. We shall examine this further and check if this order is specific to one shop or several shops.

In [28]:
df_2000 = df[df['total_items'] == 2000]
df_2000.shape[0]

17

**There are 17 transactions with `total_items` = 2000**. We now go ahead and examine if these transactions have emerged from a single shop or multiple shops. We also look at other variables like order_amount, order_id and time associated with these 17 transactions. 

In [29]:
df_2000

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
520        521       42      607        704000         2000    credit_card   
1104      1105       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
1436      1437       42      607        704000         2000    credit_card   
1562      1563       42      607        704000         2000    credit_card   
1602      1603       42      607        704000         2000    credit_card   
2153      2154       42      607        704000         2000    credit_card   
2297      2298       42      607        704000         2000    credit_card   
2835      2836       42      607        704000         2000    credit_card   
2969      2970       42      607        704000         2000    credit_card   
3332      3333       42      607        704000         2000    credit_card   
4056      4057       42      607        704000         2000    credit_card   
4646      4647       42      607        704000         2000    credit_card   
4868      4869       42      607        704000         2000    credit_card   
4882      4883       42      607        704000         2000    credit_card   

              created_at  
15    2017-03-07 4:00:00  
60    2017-03-04 4:00:00  
520   2017-03-02 4:00:00  
1104  2017-03-24 4:00:00  
1362  2017-03-15 4:00:00  
1436  2017-03-11 4:00:00  
1562  2017-03-19 4:00:00  
1602  2017-03-17 4:00:00  
2153  2017-03-12 4:00:00  
2297  2017-03-07 4:00:00  
2835  2017-03-28 4:00:00  
2969  2017-03-28 4:00:00  
3332  2017-03-24 4:00:00  
4056  2017-03-28 4:00:00  
4646  2017-03-02 4:00:00  
4868  2017-03-22 4:00:00  
4882  2017-03-25 4:00:00

Observations:

1. There is only one customer who is placing this bulk order of 2000 sneakers worth $704000.


2. This bulk order is originating from a single shop i.e. shop no. 42.


3. There are multiple orders of the same size placed on a single day at the same time, specifically on March 2nd and March 7th at 4.00 PM.


4. All these orders were placed exactly at 4 PM on different days in the month of MARCH.


5. All are credit card transactions.


6. There are different possibilities associated with this anomaly: There could either be a wrong entry, or a fraudulant transaction on a credit card. It may also be the case when a customer places bulk orders on this store and tries to resell the product in a local store at a higher price.

**Let us now examine the other orders with respect to the order size and the transaction amount.**

In [40]:
# Calculating the percentage distribution of each bundle of orders.
df['total_items'].value_counts(normalize = True)

2       0.3664
1       0.3660
3       0.1882
4       0.0586
5       0.0154
2000    0.0034
6       0.0018
8       0.0002
Name: total_items, dtype: float64

**The above distribution indicates that majority of the customers (92%) bought either 1,2, or 3 sneakers**.

In [73]:
# Looking at the minimum and maximum order amount for each bundle of sneakers bought.

df_group = df.groupby('total_items').agg({'order_amount': ['mean', 'min', 'max']}).round(0)

In [74]:
df_group

order_amount                
                    mean     min     max
total_items                             
1                  417.0      90   25725
2                  750.0     180   51450
3                 1191.0     270   77175
4                  948.0     360  102900
5                  759.0     450    1760
6                17940.0     774  154350
8                 1064.0    1064    1064
2000            704000.0  704000  704000

**When we look at the above descriptive statistics, we see that the maximum order amounts are unusually high. Let us further examine this.**

In [70]:
# Calculating maximum amount paid per sneaker.

df_group[('order_amount', 'max')]/df_group.index

total_items
1       25725.0
2       25725.0
3       25725.0
4       25725.0
5         352.0
6       25725.0
8         133.0
2000      352.0
dtype: float64

In [75]:
# Calculating minimum amount paid per sneaker.

df_group[('order_amount', 'min')]/df_group.index

total_items
1        90.0
2        90.0
3        90.0
4        90.0
5        90.0
6       129.0
8       133.0
2000    352.0
dtype: float64

In [84]:
# calculating amount paid per item

df['amount per item'] = df['order_amount']/df['total_items'] 


In [94]:
# Plotting the amount paid per item

x = df[['amount per item']]
fig = px.scatter(df, y="amount per item")
fig.show()

**As we can see from this scatterplot, most of the items are priced between `$100- $300`, however there are some priced as high as $25,725.**

This is another anomaly which we can find in the data. Since the data is highly skewed to the right due to this outlier, mean is not a correct representation of the order amount since the mean will be pulled to the direction of skewness. In such cases, median is a better represenation of data. Lets look at the median of amount per sneaker.

In [101]:
# Calculate median order amount.

df['order_amount'].median()

284.0

In [100]:
# Calculate median price per sneaker.

median_amount_per_order = df[['amount per item']].median()
      
print("Median is a better measure because: " + str(median_amount_per_order))

Median is a better measure because: amount per item    153.0
dtype: float64


# Solution to Question 1 (Summary)

### Question 1.a:

Mean is highly impacted by outliers, hence using average/mean is not a good estimate to analyse the center of the data. In this dataset, there are sneakers selling for $25725 which is significantly high than the average price of `$100-$300`. Hence mean is not an efficient way to understand the order value. 

### Question 2.a:

A better measure to evaluate skewness would be to use median. Again, this depends on the business objective. If we are looking at the average sales amount generated by the business then we need to take the median of the `order_amount` from the dataset. If we are interested to understand what is the average selling price of each sneaker then we need to calculate `order_amount/Total_items` from the dataset and estimate median of that.

### Question 3.a:

Median price per sneaker = $153.

Median order value for the business = $284.



# Solution to Question 2 (SQL)

### Question 2.a: How many orders were shipped by Speedy Express in total?

**Answer**: 54

**Code**: Select count(*)

from (select ShipperName

from Shippers as s

inner join Orders as o 

on s.ShipperID = o.ShipperID)

where ShipperName = "Speedy Express";

### Question 2.b: What is the last name of the employee with most orders?

### Answer: Peacock with 40 orders.

### Code:

select LastName, No_of_orders

from Employees

inner join

(SELECT count(*) as No_of_orders, EmployeeID

from Orders

group by EmployeeID

order by No_of_orders desc

limit 1) as new

on new.EmployeeID = Employees.EmployeeID;

### Question 2.c: What product was ordered the most by customers in Germany?
### Answer 2.c: Boston Crab Meat
### Code:


select CustomerName, Country, Total_quantity, ProductName

from Customers

inner join 

(SELECT Orders.OrderID, CustomerID, ProductID, ProductName, sum(Quantity)as Total_quantity

from Orders

inner join

(SELECT Products.ProductID, OrderID, ProductName, Quantity

from Products

inner join OrderDetails

on OrderDetails.ProductID = Products.ProductID) as new

on Orders.OrderID = new.OrderID

group by ProductName

order by Total_quantity desc) as new1

on new1.CustomerID = Customers.CustomerID

where Country == "Germany"

limit 1;
